prawnblaster/prawnblaster.cpp

```diff
@@ -1193,26 +1146,152 @@ void loop()
                     half_period += 4;
                 }
             }
-            printf("%u %u\n", half_period, reps);
+            fast_serial_printf("%u %u\r\n", half_period, reps);
+        }
+    }
+    else if (strncmp(readstring, "setb ", 5) == 0)
+    {
+		// set a large block of instructions encoded in a binary blob of fixed length.
+        unsigned int start_addr;
+        unsigned int inst_count;
+        unsigned int pseudoclock;
+        int parsed = sscanf(readstring, "%*s %u %u %u", &pseudoclock, &start_addr, &inst_count);
+        int address_offset = pseudoclock * (max_instructions * 2 / num_pseudoclocks_in_use + 2);
+        if (parsed < 3)
+        {
+            fast_serial_printf("invalid request\n");
+        }
+        else if (pseudoclock < 0 || pseudoclock > 3)
+        {
+            fast_serial_printf("The specified pseudoclock must be between 0 and 3 (inclusive)\r\n");
+        }
+        else if (start_addr >= max_instructions || start_addr + inst_count >= max_instructions)
+        {
+            fast_serial_printf("invalid address or too many instructions\r\n");
+        }
```

Comment:
I've encountered an issue here. While these checks are good, they aren't actually useful. When using this command under normal operation, you don't expect any output before writing the binary blob, so if you do hit one of these checks, you don't notice because you aren't reading from the device, and everything locks up when you send the binary blob anyway.

Example where I try to write too many instructions:
```python
instr = 40000
seq = np.array([[100, 4],[1000000, 1]], dtype=

prawn.write(f'setb 0 0 {dummy_data.shape[0]:d}\r\n'.encode())
serial_buffer = dummy_data.tobytes()

prawn.write(serial_buffer)
resp = prawn.readline().decode()
```

Presumably the check catches this, puts something in the write buffer, then execution continues,  resulting in some overflow or segfault which locks up the pico.

Now, I could put a readline before the serial buffer, but then regular operation gets slowed by the timeout since nothing is written. Defeats the purpose.

Could just error out, but then the user is all but certainly sending a giant block of data that needs to be handled regardless. So we need to read the buffer til its empty, but not save any of it. I suppose that isn't horrible. 

Anyway, we should probably fix this somehow since borking binary writes is easy, especially if we don't communicate obvious issues up front.

Reply:
Maybe the simplest option is to just return `ready\r\n` at the end of the initial checks. Users then send `setb`, check they get `ready`, then send the binary blob. If a check fails, we send the error and continue the main readline loop without executing any binary reads.